ui/job-view/context/PinnedJobs.jsx

```diff
@@ -0,0 +1,177 @@
+    };
+  }
+
+  componentDidMount() {
+    this.pinJob = this.pinJob.bind(this);
+    this.unPinJob = this.unPinJob.bind(this);
+    this.pinJobs = this.pinJobs.bind(this);
+    this.addBug = this.addBug.bind(this);
+    this.removeBug = this.removeBug.bind(this);
+    this.unPinAll = this.unPinAll.bind(this);
+    this.togglePinJob = this.togglePinJob.bind(this);
+    this.setPinBoardVisible = this.setPinBoardVisible.bind(this);
+  }
+
+  setPinBoardVisible(isPinBoardVisible) {
+    this.setState({ isPinBoardVisible });
+  }
+
+  pinJob(job, callback) {
+    const { pinnedJobs } = this.state;
+    const { notify } = this.props;
+
+    if (thPinboardMaxSize - Object.keys(pinnedJobs).length > 0) {
+      this.setState({
+        pinnedJobs: { ...pinnedJobs, [job.id]: job },
+        isPinBoardVisible: true,
+      }, () => { if (callback) callback(); });
+      this.pulsePinCount();
+    } else {
+      notify.send(thPinboardCountError, 'danger');
+    }
+  }
+
+  unPinJob(id) {
+    const { pinnedJobs } = this.state;
+
+    delete pinnedJobs[id];
+    this.setState({ pinnedJobs: { ...pinnedJobs } });
+  }
+
+  pinJobs(jobsToPin) {
+    const { pinnedJobs } = this.state;
+    const { notify, selectFirstJob } = this.props;
+    const spaceRemaining = thPinboardMaxSize - Object.keys(pinnedJobs).length;
+    const showError = jobsToPin.length > spaceRemaining;
+    const newPinnedJobs = jobsToPin.slice(0, spaceRemaining).reduce((acc, job) => ({ ...acc, [job.id]: job }), {});
+
+    if (!spaceRemaining) {
+      notify.send(thPinboardCountError, 'danger', { sticky: true });
+      return;
+    }
+
+    this.setState({
+      pinnedJobs: { ...pinnedJobs, ...newPinnedJobs },
+      isPinBoardVisible: true,
+    }, () => {
+      selectFirstJob(Object.values(newPinnedJobs));
+      if (showError) {
+        notify.send(thPinboardCountError, 'danger', { sticky: true });
+      }
+    });
+  }
+
+  addBug(bug, job) {
+    const { pinnedJobBugs } = this.state;
+
+    pinnedJobBugs[bug.id] = bug;
+    this.setState({ pinnedJobBugs: { ...pinnedJobBugs } });
+    if (job) {
+        this.pinJob(job);
+    }
+  }
+
+  removeBug(id) {
+    const { pinnedJobBugs } = this.state;
+
+    delete pinnedJobBugs[id];
+    this.setState({ pinnedJobBugs: { ...pinnedJobBugs } });
+  }
+
+  unPinAll() {
+    this.setState({
+      pinnedJobs: {},
+      pinnedJobBugs: {},
+    });
+  }
+
+  togglePinJob(job) {
+    const { pinnedJobs } = this.state;
+
+    if (pinnedJobs[job.id]) {
+      this.unPinJob(job.id);
+    } else {
+      this.pinJob(job);
+    }
+  }
+
+  pulsePinCount() {
+    const jobEl = document.getElementById('pin-count-group');
+
+    if (jobEl) {
+      jobEl.classList.add('pin-count-pulse');
+      window.setTimeout(() => {
+        jobEl.classList.remove('pin-count-pulse');
+      }, 700);
+    }
+  }
+
+  render() {
+    const { pinnedJobs, pinnedJobBugs, isPinBoardVisible } = this.state;
+
+    return (
+      <PinnedJobsContext.Provider value={{
```

Comment:
This is a really good questions.  I'm not sure yet.  :)  Experimenting now.  It DOES seem like it would apply.  But I can't be certain just yet.  Also not quite certain how to address this since there's several fields of that object that get updated independently...  :)  Might make for interesting ``setState`` calls within this class.